stl/inc/xutility

```diff
@@ -1857,6 +1802,182 @@ namespace ranges {
         inline constexpr _End::_Cpo end;
     }
 
+    template <class _Rng>
+    concept range = requires(_Rng& __r) {
+        _RANGES begin(__r);
+        _RANGES end(__r);
+    };
+
+    template <range _Rng>
+    using sentinel_t = decltype(_RANGES end(_STD declval<_Rng&>()));
+
+    // These allow us to check for unwrappability _at all_.
+    template <class _Iter>
+    concept _Weakly_unwrappable_iter = requires(_Iter&& _It) {
+        _STD forward<_Iter>(_It)._Unwrap();
+        _It._Seek_to(_STD forward<_Iter>(_It)._Unwrap());
+    };
+    template <class _Sent>
+    concept _Weakly_unwrappable_sent = requires {
+        _STD declval<_Sent>()._Unwrap();
+    };
+
+    // This is a concept since `declval<_Iter>()._Unwrapped()` not compiling will result in false,
+    // rather than a compiler error.
+    template <class _Iter>
+    concept _Nothrow_unwrappable = noexcept(_STD declval<_Iter>()._Unwrapped());
+
+    template <class _Iter>
+    decltype(auto) _Weakly_unwrap_iter(_Iter&& _It) noexcept(
+        !_Weakly_unwrappable_iter<_Iter> || _Nothrow_unwrappable<_Iter>) {
+        // This is different from _Get_unwrapped, as we would prefer pointers to be returned by reference
+        // (so as to avoid extra `seek`s).
+        if constexpr (is_array_v<remove_reference_t<_Iter>>) {
+            return _It + 0;
+        } else if constexpr (_Weakly_unwrappable_iter<_Iter>) {
+            return static_cast<_Iter&&>(_It)._Unwrapped();
+        } else {
+            return static_cast<_Iter&&>(_It);
+        }
+    }
+    template <class _Sent>
+    decltype(auto) _Weakly_unwrap_sent(_Sent&& _Se) noexcept(
+        !_Weakly_unwrappable_sent<_Sent> || _Nothrow_unwrappable<_Sent>) {
+        if constexpr (is_array_v<remove_reference_t<_Sent>>) {
+            return _Se + 0;
+        } else if constexpr (_Weakly_unwrappable_sent<_Sent>) {
+            return static_cast<_Sent&&>(_Se)._Unwrapped();
+        } else {
+            return static_cast<_Sent&&>(_Se);
+        }
+    }
+
+    template <class _Iter>
+    using _Weakly_unwrapped_iter_t = _Remove_cvref_t<decltype(_Weakly_unwrap_iter(_STD declval<_Iter>()))>;
+    template <class _Sent>
+    using _Weakly_unwrapped_sent_t = _Remove_cvref_t<decltype(_Weakly_unwrap_sent(_STD declval<_Sent>()))>;
+
+    template <class _Iter, class _Sent>
+    concept _Unwrappable_iter = _Weakly_unwrappable_iter<_Iter> //
+        && sentinel_for<_Weakly_unwrapped_sent_t<_Sent>, _Weakly_unwrapped_sent_t<_Iter>>;
+
+    template <class _Sent, class _Iter>
+    concept _Unwrappable_sent = _Weakly_unwrappable_sent<_Sent> //
+        && sentinel_for<_Weakly_unwrapped_sent_t<_Sent>, _Weakly_unwrapped_sent_t<_Iter>>;
+
+    // clang-format off
+    template <class _Sent, class _Iter>
+        requires sentinel_for<_Sent, _Iter>

```

Comment:
I hate that these are overload sets. Firstly, because I hate paying for overload resolution, and secondly because there's duck typing going on in the constraints. We can't evaluate `sentinel_for<UserRange, UserIterator>`, for example.

Maybe use `_Unwrap_range_iter`/`_Unwrap_range_sent` for the `range` flavors?